src/thunderbots/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,272 @@
+      optimizer(optimizer_param_weights),
+      best_known_pass(std::nullopt),
+      target_region(std::nullopt),
+      in_destructor(false)
+{
+    // Generate the initial set of passes
+    passes_to_optimize = generatePasses(num_passes_to_optimize.value());
+
+    // Start the thread to do the pass generation in the background
+    // The lambda expression here is needed so that we can call
+    // `continuouslyGeneratePasses()`, which is not a static function
+    pass_generation_thread =
+        std::thread([this]() { return continuouslyGeneratePasses(); });
+}
+
+void PassGenerator::setWorld(World world)
+{
+    // Take ownership of the world for the duration of this function
+    std::lock_guard<std::mutex> world_lock(world_mutex);
+
+    // Update the world
+    this->world = std::move(world);
+}
+
+void PassGenerator::setPasserPoint(Point passer_point)
+{
+    // Take ownership of the passer_point for the duration of this function
+    std::lock_guard<std::mutex> passer_point_lock(passer_point_mutex);
+
+    // Update the passer point
+    this->passer_point = passer_point;
+
+    // Replace all the passes we're currently trying to optimize, as they are probably
+    // not going to converge to this new passer point if they've already been converging
+    // to another passer point for a while
+    passes_to_optimize = generatePasses(num_passes_to_optimize.value());
+}
+
+std::optional<Pass> PassGenerator::getBestPassSoFar()
+{
+    // Take ownership of the best_known_pass for the duration of this function
+    std::lock_guard<std::mutex> best_known_pass_lock(best_known_pass_mutex);
+
+    return best_known_pass;
+}
+
+void PassGenerator::setTargetRegion(Rectangle area)
+{
+    // Take ownership of the target_region for the duration of this function
+    std::lock_guard<std::mutex> target_region_lock(target_region_mutex);
+
+    this->target_region = std::move(area);
+}
+
+
+PassGenerator::~PassGenerator()
+{
+    // Set this flag so pass_generation_thread knows to end (also making sure to
+    // properly take and give ownership of the flag)
+    in_destructor_mutex.lock();
+    in_destructor = true;
+    in_destructor_mutex.unlock();
+
+    // Join to pass_generation_thread so that we wait for it to exit before destructing
+    // the thread object. If we do not wait for thread to finish executing, it will
+    // call `std::terminate` when we deallocate the thread object and kill our whole
+    // program
+    pass_generation_thread.join();
+}
+
+void PassGenerator::continuouslyGeneratePasses()
+{
+    // Take ownership of the in_destructor flag so we can use it for the conditional
+    // check
+    in_destructor_mutex.lock();
+    while (!in_destructor)
+    {
+        // Give up ownership of the in_destructor flag now that we're done the
+        // conditional check
+        in_destructor_mutex.unlock();
+
+        optimizePasses();
+        pruneAndReplacePasses();
+        saveBestPath();
+
```

Comment:
In the old code there was logic to start from scratch if the best quality was below a certain threshold. This was useful because sometimes the pass points would behave badly when the quality was really low, so that even once the game state had improved those points were still very far from optimal

Reply:
Hmm, this is a great point. I'm going to make a note if it on the issue to implement `generatePasses` because I think it fits better in that context? (I'll be sure to tag you for review on that as well)